composable-tasks/src/compose_tasks.rs

```diff
@@ -0,0 +1,137 @@
+
+#[derive(TypeAbi, TopEncode, TopDecode, PartialEq, ManagedVecItem)]
+pub enum TaskType {
+    WrapEGLD,
+    UnwrapEGLD,
+    Swap,
+    RouterSwap,
+    SendEgldOrEsdt,
+}
+
+#[multiversx_sc::module]
+pub trait TaskCall:
+    external_sc_interactions::pair_actions::PairActionsModule
+    + external_sc_interactions::router_actions::RouterActionsModule
+    + external_sc_interactions::wegld_swap::WegldWrapModule
+{
+    #[payable("*")]
+    #[endpoint(composeTasks)]
+    fn compose_tasks(
+        &self,
+        min_expected_token_out: EgldOrEsdtTokenPayment,
+        tasks: MultiValueEncoded<MultiValue2<TaskType, ManagedVec<ManagedBuffer>>>,
+    ) {
+        let mut payment_for_next_task = self.call_value().egld_or_single_esdt();
+        let mut payments_to_return = PaymentsVec::new();
+
+        let caller = self.blockchain().get_caller();
+
+        for task in tasks.into_iter() {
+            let (task_type, args) = task.into_tuple();
+
+            let payment_for_current_task = payment_for_next_task.clone();
+
+            payment_for_next_task = match task_type {
+                TaskType::WrapEGLD => self.wrap_egld(payment_for_current_task),
+                TaskType::UnwrapEGLD => self.unwrap_egld(payment_for_current_task),
+                TaskType::Swap => {
+                    require!(
+                        !payment_for_current_task.token_identifier.is_egld(),
+                        "EGLD can't be swapped!"
+                    );
+                    let payment_in = payment_for_current_task.unwrap_esdt();
+
+                    require!(args.len() == 2, "Swap requires only 2 arguments");
+
+                    let token_out = TokenIdentifier::from(args.get(0).clone_value());
+                    let min_amount_out = BigUint::from(args.get(1).clone_value());
+
+                    self.perform_tokens_swap(
+                        payment_in.token_identifier,
+                        payment_in.amount,
+                        token_out,
+                        min_amount_out,
+                    )
+                    .into()
+                }
+                TaskType::RouterSwap => {
+                    require!(
+                        !payment_for_current_task.token_identifier.is_egld(),
+                        "EGLD can't be swapped!"
+                    );
+                    require!(
+                        args.len() % 4 == 0,
+                        "Invalid number of router swap arguments"
+                    );
+                    let payment_in = payment_for_current_task.unwrap_esdt();
+                    let mut returned_payments_by_router = self.multi_pair_swap(payment_in, args);
+
+                    let last_payment_index = returned_payments_by_router.len() - 1;
+                    let payment_out = returned_payments_by_router.take(last_payment_index);
+                    payments_to_return.append_vec(returned_payments_by_router);
```

Comment:
Just as a thought, I see why this approach is the best for building a task(to use the last payment of the vec for the next step), but would there be a better way to handle the dust payments from the router swap? Make this step even more composable?